### ISSUE TITLE
Bug with anchors and `!apply`

### ISSUE THREAD

**Issue by pplantinga** [speechbrain/HyperPyYAML#3]
This minimal example crashes and claims there's duplicate tags:

```
import hyperpyyaml

yaml_string = """
a: &id010
 - a
 - b

b: &id001 !new:collections.Counter
 - abcd

d: !apply:operator.add
 a: *id010
 b: *id001
"""

print(hyperpyyaml.load_hyperpyyaml(yaml_string))
```

gives:

```
yaml.composer.ComposerError: found duplicate anchor 'id001'; first occurrence
  in "<file>", line 1, column 4
second occurrence
  in "<file>", line 5, column 4
```

**Comment by flexthink** [comment]
It looks like the bug is triggered specifically by the presence of constructors.

This does not trigger the bug
```python
yaml_string = """
a: &id010
 - a
 - b

b: &id001 
 - abcd

d:
 a: *id010
 b: *id001
"""

print(hyperpyyaml.load_hyperpyyaml(yaml_string))
```

**Comment by flexthink** [comment]
Also it appears to be in Hyperpyyaml specifically. The rough pyyaml equivalent does not seem to be triggering the issue

```python
import yaml

class Foo:
    def __init__(self, a, b):
        self.a, self.b = a, b
    def __repr__(self):
        return f'(Foo: a={self.a}, b={self.b})'

yaml_string = """
a: &id010
 - a
 - b

b: &id001 !!python/object:__main__.Foo
 a: 1
 b: 2

d: !!python/object:__main__.Foo
 a: *id010
 b: *id001
"""

print(yaml.load(yaml_string, Loader=yaml.Loader))
```


Getting this output:
```
{'a': ['a', 'b'], 'b': (Foo: a=1, b=2), 'd': (Foo: a=['a', 'b'], b=(Foo: a=1, b=2))}
```

**Comment by flexthink** [comment]
I was able to isolate the issue to `resolve_references`

```
yaml_string = """
a: &id010
 - a
 - b

b: &id001 !new:collections.Counter
 - abcd

d: !apply:operator.add
 a: *id010
 b: *id001
"""

# %%
rref = resolve_references(
    yaml_string, None, True
)
print(rref.read())
```

The output is:

```python

rref = resolve_references(...
a: &id001
- a
- b

b: &id001 !new:collections.Counter
- abcd
d: !apply:operator.add
  a: *id001
  b: *id001
```

**Comment by flexthink** [comment]
It looks like it's ruamel.yaml that's doing this:

```python
import ruamel.yaml
ruamel_yaml = ruamel.yaml.YAML()
preview = ruamel_yaml.load(yaml_string)
yaml_stream = StringIO()
ruamel_yaml.dump(preview, yaml_stream)
yaml_stream.seek(0)
print(yaml_stream.read())
```

The output is:
```yaml
a: &id001
- a
- b

b: &id001 !new:collections.Counter
- abcd
d: !apply:operator.add
  a: *id001
  b: *id001
```

**Comment by flexthink** [comment]
Bug reported to ruamel.yaml:
https://sourceforge.net/p/ruamel-yaml/tickets/387/

**Comment by AvdN** [comment]
Yes that is caused by ruamel.yaml. It recognizes templated anchors (of the form idNNN), and doesn't set an explicit anchor if those are encountered (resulting in a potential renumbering). It forgot to check for templated anchors in undefined objects (those tagged) nor does it guarantee that all anchors are templated or all are non-templated.

This should be fixed with 0.17.8

**Comment by pplantinga** [comment]
Looks like this is solved
